### DISCUSSION TITLE
How to prevent WebSocket connection closing

### DISCUSSION THREAD

**Discussion by luciannagarro** [microsoft/vs-streamjsonrpc#990] [General]
Hello,

I have a middleware, checking for websocket request,

```
        if (context.WebSockets.IsWebSocketRequest)
        {
            using var webSocket = await context.WebSockets.AcceptWebSocketAsync();
            await using var messageHandler = new WebSocketMessageHandler(webSocket, new SystemTextJsonFormatter());
            using var jsonRpc = new JsonRpc(messageHandler, this.service);
            jsonRpc.StartListening();
            await jsonRpc.Completion;
        }
        else
        {
            await this.next(context);
        }
```

Everything works, the problem is when i send a `malformed json` it closes the connection, can i prevent closing ?

Thanks

**Comment by AArnott** [comment]
Malformed JSON will cause an exception to be thrown from the `IJsonRpcMessageFormatter`, which in turn upsets the `IJsonRpcMessageHandler`, which in turn drops the connection. 
For stream-based transports, this makes sense because message corruption cannot typically be recovered from (e.g. where in the incoming stream do you start trying to read again?).
But for WebSocket where message boundaries are built into the transport, one invalid message doesn't make finding the next message difficult. So theoretically you could build a WebSocket implementation of `IJsonRpcMessageHandler` that _doesn't_ disconnect when the formatter fails to deserialize an incoming message and instead just keeps listening.

The issue with this though, is if you can't deserialize a JSON-RPC message, you can't respond to it either to notify the sender that an error occurred. That can be worse than disconnecting, because at least in disconnecting, the sender knows something went very wrong. But the server simply dropping an invalid message without an error could leave the client hanging forever. 

There doesn't seem to be any valid cause for malformed JSON in production environments, so disconnecting seems like the right thing, IMO. And in development environments... I'm not sure what system can't reliably send valid JSON-RPC messages either, unless you're hand-typing them in, in which case disconnection seems easy enough to recover from.
